lib/pulse_counter.toit

```diff
@@ -0,0 +1,229 @@
+    ignored.
+
+  A unit is constructed in a started state. As soon as a channel is added with $add_channel it
+    starts counting.
+
+  # Advanced
+  The $glitch_filter_ns are converted to APB clock cycles. Usually the APB clock runs at 80MHz,
+    which means that the shortest glitch_filter_ns that makes sense is 1/80MHz = 12.5ns (-> 13).
+
+  The glitch filter is limited to 10 bits, and the highest value is thus 1023 ticks, or 12_787ns
+    (12.5 * 1023 = 12_787.5).
+  */
+  constructor --low/int=0 --high/int=0 --glitch_filter_ns/int?=null:
+    if glitch_filter_ns != null:
+      if glitch_filter_ns <= 0: throw "INVALID_ARGUMENT"
+      // The glitch filter runs on the APB clock (80MHz, 12.5ns), and allows at most 1023 ticks.
+      // 12.5 * 1023 == 12787.5.
+      if glitch_filter_ns > 12_787: throw "OUT_OF_RANGE"
+    else:
+      glitch_filter_ns = -1
+    unit_resource_ = pcnt_new_unit_ resource_group_ low high glitch_filter_ns
+
+  /**
+  Adds the channel to this counter.
+
+  The channel listens on the given $pin for changes. The counting mode is determined by
+    $on_positive_edge and $on_negative_edge. These parameters must be one of:
+  - $DO_NOTHING
+  - $INCREMENT
+  - $DECREMENT
+
+  The $control_pin can be used to change the mode of operation of the channel. The $when_control_low
+    and $when_control_high parameters must be one of:
+  - $KEEP: The $control_pin does not affect the mode of operation for the selected state.
+  - $REVERSE: The $control_pin reverses the effect of the mode of operation. If the channel's pin
+    was incrementing on an edge, it now decrements. If it was decrementing it now increments.
+  - $DISABLE: The $control_pin disables the channel. No changes to the unit's $value happen when
+    this control mode is active.
+  */
+  add_channel pin/gpio.Pin -> Channel
+      --on_positive_edge /int = INCREMENT
+      --on_negative_edge /int = DO_NOTHING
+      --control_pin /gpio.Pin? = null
+      --when_control_low /int = KEEP
+      --when_control_high /int = KEEP:
+     return Channel.private_ this pin \
+         on_positive_edge on_negative_edge \
+         control_pin when_control_low when_control_high
+
+  /**
+  Closes this unit.
+
+  Frees all the underlying resources.
+  */
+  close:
+    if closed_: return
+    // TODO(florian): do we need to keep track of the channels and close them?
+    // The underlying resources are already freed.
+    closed_ = true
+    pcnt_close_unit_ unit_resource_
+
+  /**
+  The value of the counter.
+
+  The ESP32 hardware supports up to 16 bits, but the range can be limited by providing
+    `high` and `low` values to the constructor.
+  */
+  value -> int:
+    return pcnt_get_count_ unit_resource_
+
+  /**
+  Resets the counter to 0.
+  */
+  clear -> none:
+    pcnt_clear_ unit_resource_
+
+  /**
+  Resumes the counter.
+
+  It is safe to call this method multiple times.
+  */
+  start -> none:
+    pcnt_start_ unit_resource_
```

Comment:
What happens if you call this methods after closing?

Reply:
added checks.